### PR TITLE
Dropout Labs Inc., Contributions

### LINKED ISSUE #17: Finish conv2d
Some combinations for Pond's `conv2d` is missing.

### LINKED ISSUE #21: Unit testing framework
Make it easy to add and run unit tests; introduce some as well.

### LINKED ISSUE #22: Easy deployment to AWS/GCP
Make it easy to quickly get up and running on AWS or GCP. Could take the form of documentation or code (perhaps along [these lines](https://github.com/mortendahl/privateml/tree/master/tensorflow/spdz/configs/gcp))

### LINKED ISSUE #25: Setup linting
We should add a linter to the code base so our code is conformant!

### LINKED ISSUE #26: Load a pretrained model for prediction
Details to come

### LINKED ISSUE #32: Train suite of models with Tensorflow to be benchmarked

### LINKED ISSUE #52: New Benchmark using tfe
Goals:
- Benchmarking tfe computation time locally with 3 players using a  5 layers conv2d + sigmoid models
- Benchmarking tfe computation time remotely with 3 players on AWS around North America with 5 layers conv2d + sigmoid

### LINKED ISSUE #55: Flatten operation import conversion
To be able import a flatten operation we need to be able to run the following operations on the underlying tensors:

- [x] Pack
- [x] Reshape
- [x] StridedSlice
- [x] Shape

### LINKED ISSUE #63: Add flag to specify enabling/disabling debug output
This way we don't write output unless we need to!

### PR DESCRIPTION
**Pull request by ianlivingstone**
This pull request contains work done by Dropout Labs Inc. over the last two weeks in a private fork. 

We're contributing it back to the community project and will continue to contribute moving forward directly to this repository. 

Closes #17 
Closes #21 
Closes #22 
Closes #25 
Closes #26 
Closes #32 
Closes #52 
Closes #55 
Closes #63

### CODE REVIEW

File: tensorflow_encrypted/config.py [hunk 1]
```diff
@@ -23,16 +35,15 @@ class LocalConfig(Config):
...
     """
 
-    def __init__(self, player_names:List[str], job_name:str='localhost') -> None:
+    def __init__(self, player_names: List[str], job_name: str='localhost') -> None:
         self._job_name = job_name
         self._players = {
             name: Player(
                 name=name,
-                index=index+1,
+                index=index,
```
Comment (mortendahl) [inline tensorflow_encrypted/config.py]: need to decide what is right thing to do here

File: tensorflow_encrypted/config.py [hunk 2]
```diff
@@ -119,14 +136,14 @@ def __init__(self,
         self._players = {
             name: Player(
                 name=name,
-                index=index+self._offset,
+                index=index,
```
Comment (mortendahl) [inline tensorflow_encrypted/config.py]: same here

File: tensorflow_encrypted/config.py [hunk 3]
```diff
@@ -154,35 +171,45 @@ def to_dict(self) -> Dict:
...
+    def session(self, log_device_placement: bool = False) -> tf.Session:
+        global __TFE_DEBUG__
 
-    def session(self, log_device_placement:bool=False) -> tf.Session:
         config = tf.ConfigProto(
             log_device_placement=log_device_placement,
             allow_soft_placement=False,
         )
-        return tf.Session(
+        print('Starting session on target: %s' % self._target, config)
```
Comment (mortendahl) [inline tensorflow_encrypted/config.py]: are all these only meant for debugging right now or would we like them to stay?

File: tensorflow_encrypted/convert/convert.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+class ConvertInputProvider(InputProvider):
+    input: tf.Tensor
+
+    def __init__(self, player: Player, input: tf.Tensor) -> None:
+        self.input = input
+        self.player = player
+
+    def provide_input(self) -> tf.Tensor:
+        return tf.identity(self.input)
```
Comment (mortendahl) [inline tensorflow_encrypted/convert/convert.py]: why do we need the `tf.identity` here?

File: tensorflow_encrypted/convert/convert.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+
+class Converter():
+    config: Config
+    protocol: Pond
```
Comment (mortendahl) [inline tensorflow_encrypted/convert/convert.py]: should probably be `Protocol` instead of `Pond`

File: tensorflow_encrypted/layers/core.py [hunk 6]
```diff
@@ -1,24 +1,37 @@
-from typing import Optional
+from typing import Optional, List
+
 from abc import ABC, abstractmethod
 from ..protocol.protocol import get_protocol, Protocol
-
-from ..protocol.pond import PondPrivateTensor
+from ..protocol.pond import TFEVariable
```
Comment (mortendahl) [inline tensorflow_encrypted/layers/core.py]: we should avoid the dependency on `pond` here

File: tensorflow_encrypted/tensor/int100.py [hunk 7]
```diff
@@ -198,9 +216,9 @@ def _sample_uniform(shape):
     return Int100Tensor.from_decomposed(backing)
 
 
-def _transpose(x, *axes):
+def _transpose(x, perm=None):
     assert isinstance(x, Int100Tensor), type(x)
-    backing = [tf.transpose(xi, axes) for xi in x.backing]
+    backing = [tf.transpose(xi, perm=perm) for xi in x.backing]
```
Comment (mortendahl) [inline tensorflow_encrypted/tensor/int100.py]: the iteration over x.backing should be moved to crt

File: tensorflow_encrypted/tensor/int100.py [hunk 8]
```diff
@@ -210,12 +228,24 @@ def _strided_slice(x: Int100Tensor, args: Any, kwargs: Any):
     return Int100Tensor.from_decomposed(backing)
 
 
-def _reshape(x, *axes):
+def _reshape(x: Int100Tensor, axes: List[int]) -> Int100Tensor:
     assert isinstance(x, Int100Tensor), type(x)
     backing = [tf.reshape(xi, axes) for xi in x.backing]
```
Comment (mortendahl) [inline tensorflow_encrypted/tensor/int100.py]: iteration should happen in crt

File: tensorflow_encrypted/tensor/int100.py [hunk 9]
```diff
@@ -210,12 +228,24 @@ def _strided_slice(x: Int100Tensor, args: Any, kwargs: Any):
...
     return Int100Tensor.from_decomposed(backing)
 
 
+def _expand_dims(x, axis=None):
+    assert isinstance(x, Int100Tensor), type(x)
+    backing = [tf.expand_dims(xi, axis) for xi in x.backing]
```
Comment (mortendahl) [inline tensorflow_encrypted/tensor/int100.py]: iteration in crt

File: tensorflow_encrypted/tensor/int100.py [hunk 10]
```diff
@@ -210,12 +228,24 @@ def _strided_slice(x: Int100Tensor, args: Any, kwargs: Any):
...
+    return Int100Tensor.from_decomposed(backing)
+
+
+def _squeeze(x, axis=None):
+    assert isinstance(x, Int100Tensor), type(x)
+    backing = [tf.squeeze(xi, axis=axis) for xi in x.backing]
```
Comment (mortendahl) [inline tensorflow_encrypted/tensor/int100.py]: iteration in crt

File: tensorflow_encrypted/tensor/int100.py [hunk 11]
```diff
@@ -236,7 +266,7 @@ def __init__(self, native_value: np.ndarray, int100_value=None) -> None:
 
         assert type(int100_value) in [Int100Tensor], type(int100_value)
 
-        backing = [tf.convert_to_tensor(vi, dtype=INT_TYPE) for vi in int100_value.backing]
+        backing = [tf.constant(vi, dtype=INT_TYPE) for vi in int100_value.backing]
```
Comment (mortendahl) [inline tensorflow_encrypted/tensor/int100.py]: iteration in crt

File: tensorflow_encrypted/config.py [hunk 12]
```diff
@@ -83,6 +95,11 @@ def session(self, log_device_placement:bool=False) -> tf.Session:
                 intra_op_parallelism_threads=1
             )
         )
+        if __TFE_DEBUG__:
+            print('session in debug mode')
```
Comment (mortendahl) [inline tensorflow_encrypted/config.py]: debug print?

File: tensorflow_encrypted/protocol/__init__.py [hunk 13]
```diff
@@ -2,14 +2,17 @@
 
 # from unencrypted_native import UnencryptedNative
 # from unencrypted_fixedpoint import UnencryptedFixedpoint
-from .pond import Pond
+from .pond import Pond, TFEVariable, TFETensor
```
Comment (mortendahl) [inline tensorflow_encrypted/protocol/__init__.py]: the general types should be moved out of Pond

Comment (morgangiraud) [reply tensorflow_encrypted/protocol/__init__.py]: I've added them back because I was running into circular dependencies.

If you create a `types` file to define Union of types like `TFEVariable` depending on classes that are inside of another file and the class have functions that you want to type with the types in the `type` file. 
Python is not happy with it.

This probably means that we too much things inside this `Pond` file.

File: tensorflow_encrypted/config.py (hunk 12 above)
Comment (morgangiraud) [reply tensorflow_encrypted/config.py]: ?

File: tensorflow_encrypted/tensor/int100.py (hunk 11 above)
Comment (mortendahl) [reply tensorflow_encrypted/tensor/int100.py]: #92

File: tensorflow_encrypted/tensor/int100.py (hunk 10 above)
Comment (mortendahl) [reply tensorflow_encrypted/tensor/int100.py]: #92

File: tensorflow_encrypted/tensor/int100.py (hunk 9 above)
Comment (mortendahl) [reply tensorflow_encrypted/tensor/int100.py]: #92

File: tensorflow_encrypted/tensor/int100.py (hunk 7 above)
Comment (mortendahl) [reply tensorflow_encrypted/tensor/int100.py]: #92

File: tensorflow_encrypted/tensor/int100.py (hunk 8 above)
Comment (mortendahl) [reply tensorflow_encrypted/tensor/int100.py]: #92

File: tensorflow_encrypted/config.py (hunk 12 above)
Comment (mortendahl) [reply tensorflow_encrypted/config.py]: is it printing just for debugging purposes or should it be kept around?

File: tensorflow_encrypted/protocol/__init__.py (hunk 13 above)
Comment (mortendahl) [reply tensorflow_encrypted/protocol/__init__.py]: maybe we need to introduce some base tensor types that the pond classes inherit from

File: tensorflow_encrypted/convert/convert.py (hunk 5 above)
Comment (mortendahl) [reply tensorflow_encrypted/convert/convert.py]: #93

File: tensorflow_encrypted/config.py (hunk 3 above)
Comment (morgangiraud) [reply tensorflow_encrypted/config.py]: I added this one for debugging at first, but I find it quite useful overall in the end. It feels like when someone starts a web server.

File: tensorflow_encrypted/layers/core.py (hunk 6 above)
Comment (mortendahl) [reply tensorflow_encrypted/layers/core.py]: #93

File: tensorflow_encrypted/protocol/__init__.py (hunk 13 above)
Comment (mortendahl) [reply tensorflow_encrypted/protocol/__init__.py]: #93

File: tensorflow_encrypted/config.py (hunk 1 above)
Comment (mortendahl) [reply tensorflow_encrypted/config.py]: #94

File: tensorflow_encrypted/config.py (hunk 2 above)
Comment (mortendahl) [reply tensorflow_encrypted/config.py]: #94
